src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java

```diff
@@ -872,7 +872,7 @@ public void readFrom(StreamInput in) throws IOException {
         @Override
         public void writeTo(StreamOutput out) throws IOException {
             super.writeTo(out);
-            final File[] files = fileDetails.values().toArray(new File[0]);
+            final File[] files = fileDetails().toArray(new File[0]);
```

Comment:
GOOD CATCH. I think the right solution is to synchronize the method , just as all other methods are. Can you also do it to the readFrom?